src/common/helpers/poller.js

```diff
@@ -0,0 +1,272 @@
+  const repo_url = getGitHubRepoUrl(owner, name);
+  if (await hasRepoChanged(repo_url, last_polled_at, token)) {
+    logger.info(`The ${owner}/${name} repository has changed.`);
+    return true;
+  }
+  logger.info(`${owner}/${name}: unchanged, checking parts ...`);
+
+  let snapcraft_yaml;
+  try {
+    snapcraft_yaml = await internalGetSnapcraftYaml(owner, name, token);
+  } catch (e) {
+    return false;
+  }
+  for (const source_part of extractPartsToPoll(snapcraft_yaml.contents)) {
+    logger.info(`${owner}/${name}: Checking whether ${source_part.repoUrl} part has changed.`);
+    if (await source_part.hasRepoChangedSince(last_polled_at, token)) {
+      logger.info(`${owner}/${name}: ${source_part.repoUrl} changed.`);
+      return true;
+    }
+  }
+  return false;
+};
+
+
+// Request a build of a given snap repository (in LP) and update `polled_at` (in DB).
+export const buildSnapRepository = async (owner, name) => {
+  const _now = new Date().getTime();
+  // XXX request LP build
+  // const repo_url = getGitHubRepoUrl(owner, name);
+  return db.transaction(async (trx) => {
+    const row = await db.model('Repository')
+      .where({ owner, name })
+      .fetch({ transacting: trx });
+    await row.save({ polled_at: _now }, { method: 'update', transacting: trx });
+  });
+};
+
+
+// Whether a given (GitHub) repository has new commits since 'last_polled_at'.
+export const hasRepoChanged = async (repositoryUrl, last_polled_at, token) => {
+  if (last_polled_at === undefined || !last_polled_at) {
+    throw new Error('`last_polled_at` must be given.');
+  }
+  const last_polled = moment(last_polled_at);
+  const since = last_polled.toISOString();
+  const { owner, name } = parseGitHubRepoUrl(repositoryUrl);
+  const uri = `/repos/${owner}/${name}/commits?since=${since}`;
+  const options = {
+    token,
+    headers: {
+      'If-Modified-Since': last_polled.format('ddd, MM MMM YYYY HH:mm:ss [GMT]')
+    },
+    json: true
+  };
+
+  const response = await requestGitHub.get(uri, options);
+
+  switch (response.statusCode) {
+    case 200:
+      // If the (JSON encoded) body is not an empty list.
+      return response.body.length > 0;
+    case 304:
+      // `If-Modified-Since` in action, cache hit, no changes.
+      return false;
+    default:
+      // Bail, unexpected response.
+      throw new Error(
+        `${repositoryUrl} (${response.statusCode}): ${response.body.message}`);
+  }
+};
+
+
+// Extracts unique GH repository URLs from a given (parsed) snapcraft.yaml
+export function extractPartsToPoll(snapcraft_yaml) {
+  const parts = Object.values(snapcraft_yaml.parts || {});
+  const sourceParts = parts.map(
+    GitSourcePart.fromSnapcraftPart).filter(part => part != undefined);
+  return Array.from(new Set(sourceParts));
+}
+
+
+/// GitSourcePart encapsulates the relevant information from a snapcraft
+/// source part, and contains the methods to determine whether it's up to
+/// date.
+export class GitSourcePart {
```

Comment:
Quite a lot of code in this file, maybe later it can be split into several files with at least this class being a separate module.